src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.h

```diff
@@ -0,0 +1,962 @@
+  }
+  // no object likes each other at this level
+  constexpr likes_t likes(const fixed_key_3&) const {
+    return likes_t::no;
+  }
+  void update_with_oid(const ghobject_t& oid) {
+    snap = oid.hobj.snap;
+    gen = oid.generation;
+  }
+  void update_oid(ghobject_t& oid) const {
+    oid.hobj.snap = snap;
+    oid.generation = gen;
+  }
+};
+
+static inline std::ostream& operator<<(std::ostream& os, const fixed_key_3& k) {
+  if (k.snap != CEPH_NOSNAP) {
+    os << "s" << k.snap << ",";
+  }
+  if (k.gen != ghobject_t::NO_GEN) {
+    os << "g" << k.gen << ",";
+  }
+  return os;
+}
+
+// all elements in this node must share the same <shard, pool, hash, namespace, oid>
+// but the unlike other FixedKey<>, a node with FixedKey<3> does not have
+// variable_sized_key, so if it is an inner node, we can just embed the child
+// addr right in the key.
+template<>
+struct FixedKey<3, ntype_t::inner> : public fixed_key_3 {
+  // the item is embedded in the key
+  static constexpr bool item_in_key = true;
+  laddr_t child_addr = 0;
+
+  FixedKey() = default;
+  void set(const ghobject_t& oid, laddr_t new_child_addr) {
+    update_with_oid(oid);
+    child_addr = new_child_addr;
+  }
+  // unlikely get called, though..
+  void update(const ghobject_t& oid) {}
+  template<int N>
+  std::enable_if_t<N < 3> set(const FixedKey<N, ntype_t::inner>&,
+                              laddr_t new_child_addr) {
+    child_addr = new_child_addr;
+  }
+  void set(const FixedKey& k, laddr_t new_child_addr) {
+    snap = k.snap;
+    gen = k.gen;
+    child_addr = new_child_addr;
+  }
+  void set(const variable_sized_key_t& k, laddr_t new_child_addr) {
+    snap = k.snap;
+    gen = k.gen;
+    child_addr = new_child_addr;
+  }
+};
+
+template<>
+struct FixedKey<3, ntype_t::leaf> : public fixed_key_3 {
+  static constexpr bool item_in_key = false;
+  uint16_t offset = 0;
+
+  FixedKey() = default;
+  void set(const ghobject_t& oid, uint16_t new_offset) {
+    update_with_oid(oid);
+    offset = new_offset;
+  }
+  void set(const FixedKey& k, uint16_t new_offset) {
+    snap = k.snap;
+    gen = k.gen;
+    offset = new_offset;
+  }
+  template<int N>
+  std::enable_if_t<N < 3> set(const FixedKey<N, ntype_t::leaf>&,
+                              uint16_t new_offset) {
+    offset = new_offset;
+  }
+};
+
+// prefix-compress the key
+struct prefix_compressed_key_t {
+  struct index_t {
+    enum {
+      shared_key_len,
+      unshared_key_len,
+      onode_len,
+      key_data,
+      onode_data,
+    };
+  };
+  using layout_type = asci::Layout<uint8_t,
+                                   uint8_t,
+                                   uint16_t,
+                                   char,
+                                   char>;
+  constexpr static layout_type cell_layout(uint8_t unshared_len,
+                                           uint16_t onode_len) {
+    return layout_type{1, 1, 1, unshared_len, onode_len};
+  }
+};
```

Comment:
Maybe it's better to include this in the future PR with prefix compression related implementations?
I don't see any users in this PR and not sure about the specific meanings.

Reply:
will do.